app/models/filter_updater.rb

```diff
@@ -0,0 +1,210 @@
+    create_multiple(item_id, missing_inherited, true)
+  end
+
+  # Notify the filterable class about the changes that we made, so that it can
+  # perform the appropriate steps to reindex everything.
+  def reindex_changed
+    klass.reindex_for_filter_changes(@valid_item_ids, @modified, queue)
+  end
+
+  ########################################
+  # RETRIEVE INFO FROM DATABASE
+  ########################################
+
+  # Calculates @direct_filters, @meta_tags, and @inherited_filters for this
+  # batch of items.
+  def load_info
+    load_valid_item_ids
+    load_direct_filters
+    load_meta_tags
+    load_inherited_filters
+  end
+
+  # Calculate which items exist in the database, so that we don't try to create
+  # FilterTaggings for items that have been deleted.
+  def load_valid_item_ids
+    @valid_item_ids = klass.unscoped.where(id: ids).distinct.pluck(:id)
+  end
+
+  # Calculates what the direct filters should be for this batch of items.
+  #
+  # Sets @direct_filters equal to a hash mapping from item IDs to a list of
+  # direct filter IDs (that is, filters that the item is either directly tagged
+  # with, or tagged with one of its synonyms). The default value for the
+  # hash is an empty list.
+  def load_direct_filters
+    taggings = Tagging.where(taggable_type: type, taggable_id: @valid_item_ids)
+
+    filter_relations = [
+      Tag.canonical.joins(:taggings),
+      Tag.canonical.joins(mergers: :taggings)
+    ]
+
+    pairs = filter_relations.flat_map do |filters|
+      filters.merge(taggings).pluck("taggings.taggable_id", "tags.id")
+    end
+
+    @direct_filters = hash_from_pairs(pairs)
+  end
+
+  # Reads MetaTagging info from the database for all tags included in this
+  # batch.
+  #
+  # Sets @meta_tags equal to a hash mapping from tag IDs to the tag's meta tag
+  # IDs. The default value for the hash is an empty list.
+  def load_meta_tags
+    all_filters = @direct_filters.values.flatten.uniq
+
+    pairs = Tag.canonical.joins(:sub_taggings).where(
+      meta_taggings: { sub_tag_id: all_filters }
+    ).pluck(:sub_tag_id, :meta_tag_id)
+
+    @meta_tags = hash_from_pairs(pairs)
+  end
+
+  # Uses @direct_filters and @meta_tags to calculate what the inherited filters
+  # should be for each of the items in this batch. Creates a hash
+  # @inherited_filters mapping from item IDs to the inherited tag IDs.
+  def load_inherited_filters
+    @inherited_filters = Hash.new([].freeze)
+
+    @direct_filters.each_pair do |item_id, filter_ids|
+      inherited = filter_ids.flat_map { |filter_id| @meta_tags[filter_id] }
+      @inherited_filters[item_id] = (inherited - filter_ids).uniq
+    end
+  end
+
+  # Given a list of pairs of IDs, treat each pair as a (key, value) pair, and
+  # return a hash that associates each key with a list of values. Sets the
+  # default value of the hash to an empty frozen list.
+  def hash_from_pairs(pairs)
+    hash = Hash.new([].freeze)
```

Comment:
I froze the default value because having a non-frozen default value can cause confusion if you ever try to modify the default value. All of the other lists in the hash shouldn't have that issue, but I can still freeze them if you think it'd be useful.

Reply:
I think we should. Freezing will make it clear we don't expect to modify the lists afterwards.